### ISSUE TITLE
How to evaluate the delta expression for convolution?

### ISSUE THREAD

**Issue by Mikolaj** [Mikolaj/horde-ad#24]
E.g., of convolution such as this 

https://hackage.haskell.org/package/hmatrix-0.20.2/docs/Numeric-LinearAlgebra.html#v:conv2

but also of `conv`, `corr2`, `corr` from the same API. BTW, it's possible the derivative of `conv2` can be defined in terms of these other operations (not of their derivatives), because gradient definitions seems to involve them to some extend (I haven't found a good source for that yet, though, and APIs and presentations vary, which adds to the confusion). BTW, it's really hard to google derivative of convolution and much easier to google gradient, which indicates that our approach of computing gradients via derivatives may be indeed unique.

This is needed for #9, because even if we can define convolution on dual numbers using primitive operations on dual numbers [edit: pushed to master branch; not tested yet], it's going to be slow, both to compute the value and especially the gradient. It's also not clear we are going to add enough primitive operations to make this possible. Therefore, a fused derivative for the whole convolution seems mandatory and a convolution definition in terms of primitive operations is a luxury that would be used mostly to QuickCheck that the fused derivative is correct (we would verify correctness separately for derivative and gradient calculations, because both work now).

Edit: or rather the im2col trick is the mandatory part and everything else is a luxury, in particular, because the hmatrix.conv2 seems to not be very optimized; see also https://www.reddit.com/r/haskell/comments/sv172y/rank_3_stencils_for_efficient_parallel_stencil/hxese33/?utm_source=reddit&utm_medium=web2x&context=3

**Comment by tomjaguarpaw** [comment]
I'm not sure what you mean by "derivative" as distinct from "gradient". Anyway,

```
conv (x + dx) (y + dx) ~= conv x y + conv x dy + conv dx y
```

(for small values of `dx` and `dy`). Does that help answer the question? If not could you say exactly what you mean by "derivative"?

**Comment by Mikolaj** [comment]
Thank you for the tip. It's not yet in a form that I could immediately test. What I need for #9 is the ??? part in

```Haskell
conv2' :: DualMonad r m
       => DualNumber (Tensor2 r) -> DualNumber (Tensor2 r)
       -> m (DualNumber (Tensor2 r))
conv2' (D u u') (D v v') = D (HM.conv2 u v) ???
```

and similarly for the other operations (starting from the 1-dimensional ops is going to be easier).

The ??? thing is going to looks very similar to a mathematical expresison for a derivative (not gradient).

Regarding how "derivative" is distinct from "gradient", note that a derivative of a function f :: R^5 -> R at point x0 :: R^5 is a function from R^5 to R, while gradient of f at x0 is a function from R to R^5. However, given that these functions are linear, you can compute one from another, as you wrote in your paper. But the library (following the paper) fixes what is computed from what, so we want a delta-expression that's going to look similar to a mathematical formula for derivative, not gradient. See the screenshot in https://github.com/Mikolaj/mostly-harmless/discussions/19 for how the two formulas can differ.

**Comment by tomjaguarpaw** [comment]
I see. Then I think

```
??? = conv2D u v' + conv2D u' v
```

where `conv2D` is the "symbolic" constructor of `Delta` corresponding to `conv2`. Does that make sense?

**Comment by tomjaguarpaw** [comment]
... and in fact rather than `... + ...` I should have written `Add ... ...`, because the addition should be done "symbolically" in `Delta` too.

**Comment by tomjaguarpaw** [comment]
By analogy with [your example of dot product](https://github.com/Mikolaj/horde-ad/blob/1e834609b973c6ad67dc4c5b269666939ef3e824/src/HordeAd/Core/DualNumber.hs#L147-L150=), I guess you would write

```haskell
conv2' :: DualMonad r m
       => DualNumber (Tensor2 r) -> DualNumber (Tensor2 r)
       -> m (DualNumber (Tensor2 r))
conv2' (D u u') (D v v') =
  D (HM.conv2 u v) (dAdd (dConv2 u v') (dConv2 u' v)
```

but I suppose that means you have to add `dConv2` to `HasRanks`, so I don't know if that helps you or just raises further questions ...

**Comment by Mikolaj** [comment]
Yes, this looks good. If that's true, that solves half of the problem. The other half is either extending `eval2` for the case of the new `Conv2` constructor or using more primitive operations (surely some of them new) instead of `dConv2`.

However, there's this typing problem. Dual number are

    data DualNumber a = D (Primal a) a

so the first usage of `dConv2` is at `(Primal a) -> a -> a` and the second is at `a -> (Primal a) -> a`, where `a = Tensor2 r`.

**Comment by tomjaguarpaw** [comment]
> However, there's this typing problem.

Hmm yes, I see. But why does the same problem not arise for dot?

> The other half is either extending eval2 for the case of the new Conv2

Sure, but don't you already know the answer to that because you were able to successfully Google for "gradient of convolution"?

**Comment by Mikolaj** [comment]
> > However, there's this typing problem.
> 
> Hmm yes, I see. But why does the same problem not arise for dot?

That's because dot is luckily commutative. However, matrix multiplication is not, so probably this could be made to type-check similarly as the terrible `dMD_M2` & Co (but with a better name?). But surely they would not evaluate the same as `dMD_M2`.

> > The other half is either extending eval2 for the case of the new Conv2
> 
> Sure, but don't you already know the answer to that because you were able to successfully Google for "gradient of convolution"?

Firstly, gradient is not good enough and the scarce google hits for "derivative of convolution" tend to be about physics and so don't quite match (different primitive operations, different concept of convolution, etc.). Secondly, knowing the derivative in maths terms would help more with "using more primitive operations (surely some of them new) instead of dConv2", than with "extending eval2 for the case of the new Conv2". OTOH, all this is related so understanding one would help with the other.

**Comment by tomjaguarpaw** [comment]
> That's because dot is luckily commutative.

Aha, I wondered why the argument order for dot was reversed in one of the summands.

> gradient is not good enough

I'm not sure what you mean. `eval2` *is* calculating the gradient.

**Comment by Mikolaj** [comment]
> I'm not sure what you mean. `eval2` _is_ calculating the gradient.

`buildFinMaps.eval2` is, but indirectly, via delta expressions. OTOH `evalBindingsForward.eval2` is calculating the derivative via a delta expression and notice that it's more straightforward. Similarly, the delta expressions, IMHO, are much closer to mathematical expressions for derivatives than gradients for the same objective function. In particular, I find it much harder to derive a delta expression for a dual number for an objective function from a corresponding mathematical definition of a gradient than of a derivative.

**Comment by tomjaguarpaw** [comment]
OK, so what are we stuck on here? I've lost track.

**Comment by Mikolaj** [comment]
> The other half is either extending eval2 for the case of the new Conv2 constructor or using more primitive operations (surely some of them new) instead of dConv2.

That would now be "two Conv2 constructors". BTW, I'm going to push

> define convolution on dual numbers using primitive operations on dual numbers

in a moment which, in particular, adds a few more primitive operations (including delta-expression constructors) that may prove useful. Also, with `conv2` definition in terms of other dual number operations, the task "using more primitive operations (surely some of them new) instead of dConv2" boils down to "apply fusion to `conv2`", see the 'fusion' mentioned in the paper. This is by no means trivial in practice, because the expression is going to be enormous if generated via brute force from the fusion formula in the paper (and the formula is only for scalars, in some different formalism, etc.).

**Comment by Mikolaj** [comment]
Also, what I'm going to push is certainly buggy, so the fusion may involve fixing it.

**Comment by Mikolaj** [comment]
Pushed. Now it will take me some days to test the naive `conv2` by defining and training a convolutional neural network for MNIST, which will be particularly hard, because training will take forever, so debugging is going to be slow.

`conv2` with a fused gradient/derivative should be much faster.

The im2col trick would probably be even faster and may or may not directly reuse bits from the naive and the fused approach. E.g., it may have exactly the same fused derivative, but it's also possible the fused derivative can be evaluated faster if reformulated according to the im2col idea. In hindsight it's surely going to look obvious. :)

**Comment by tomjaguarpaw** [comment]
Here are some observations that will prove useful regardless of the problem at hand.

Let `conv` be the convolution function on (bi)infinite 1-d vectors (as such it only makes sense mathematically, not computationally)

```
conv(x, y)_k = sum_{i + j = k} x_i y_j
```

Now let's consider the dot product of `conv(x, y)` and `z`

```
conv(x, y) . z
= sum_k conv(x, y)_k z_k
= sum_k sum_{i + j = k} x_i y_j z_k
= sum_{i, j} x_y y_j z_{i + j}
```

and the dot product of `x` and `conv(y, z)`

```
x . conv(y, z)
= sum_k sum_{i + j = k} x_k y_i z_j
= sum_{i, j} x_{i + j} y_i z_j
```

The two summations are the same, just with the indices rearranged, so we conclude that

```
conv(x, y) . z == x . conv(y, z)
```

Here's another observation

```
conv(x, y) = sum_{i + j = k} x_i y_j
conv(y, y) = sum_{i + j = k} x_j y_i
```

Again the two summations are the same, with indices rearranged., So

```
conv(x, y) == conv(y, x)
```

We can combine the observations with our earlier observation about the derivative of `conv`, that is

```
dconv(x, y) = conv(x, dy) + conv(dx, y)
```

Let `r = conv(x, y)`. Then the defining property of the gradient `∂ / ∂r` is that

```
∂ / ∂r . dr = ∂ / ∂x dx + ∂ / ∂y dy
```

So we are looking for `∂ / ∂x` and `∂ / ∂y` that solve the equation

```
∂ / ∂r . (conv(x, dy) + conv(dx, y)) = ∂ / ∂x . dx + ∂ / ∂y . dy
```

By separating the dependency on `dx` and `dy` we are looking for solutions to the following two equations

```
∂ / ∂r . conv(x, dy) = ∂ / ∂y . dy
∂ / ∂r . conv(dx, y) = ∂ / ∂x . dx
```

By the property above

```
∂ / ∂r . conv(x, dy) = conv(∂ / ∂r, x) . dy
∂ / ∂r . conv(dx, y) = ∂ / ∂r . conv(y, dx) = conv(∂ / ∂r, y) . dx
```

so the solutions to the defining equation are

```
∂ / ∂y = conv(∂ / ∂r, x)
∂ / ∂x = conv(∂ / ∂r, y) = conv(y, ∂ / ∂r)
```

i.e. this is how `conv` acts on gradients.

Now, there are a few things to bear in mind converting this to a practical use case

1. This is for 1-d *infinite* vectors. The same reasoning should hold for n-d infinite vectors.
2. For n-d *finite* vectors the same reasoning should hold when `conv` is interpreted as convolution-with-zero-padding.
3. It's typically left ambiguous, particularly in informal discussions, whether "correlation" means `conv` as I have used above or "correlation", that is `corr(x, y) = conv(x^R, y)` where `x^R` is `x` rotated through 180 degrees. If so then the gradient calculation will have to be adjusted accordingly (cf. [a Quora answer](https://www.quora.com/How-does-one-calculate-the-gradients-of-a-convolution-neural-network)).

**Comment by tomjaguarpaw** [comment]
> That's because dot is luckily commutative

[Convolution is also luckily communative](https://en.wikipedia.org/wiki/Convolution#Properties) (assuming it actually means convolution, and not "correlation" as some people mean).  [Hmatrix convolutions](https://hackage.haskell.org/package/hmatrix-0.20.2/docs/Numeric-LinearAlgebra.html#v:conv2) do seem to mean actual convolution.

**Comment by tomjaguarpaw** [comment]
```
> let x = fromLists [[1,2], [3,4]]
> let y = fromLists [[5, 6, 7], [8, 9, 10]]
> conv2 x y :: Matrix Double
(3><4)
 [  5.0, 16.0, 19.0, 14.0
 , 23.0, 63.0, 73.0, 48.0
 , 24.0, 59.0, 66.0, 40.0 ]
> conv2 y x :: Matrix Double
(3><4)
 [  5.0, 16.0, 19.0, 14.0
 , 23.0, 63.0, 73.0, 48.0
 , 24.0, 59.0, 66.0, 40.0 ]
```

**Comment by Mikolaj** [comment]
Hey, thanks for the observations. While I can't easily relate the infinite case to the ML convolution, it confirms my impression that the derivative of convolution is itself a convolution or somewhere close. It follows that if we speed up computing convolution (via one of the steps outlined earlier), that is likely to also speed up computing the gradient of convolution, which is the real bottleneck. So it's worth trying.

Regarding commutativity, that's also great news and I would never guess that. Apparently, through padding, correlation can be made commutative (`conv2` uses `corr2` in a rather simple way, except for the padding; see rather the hmatrix code than our code, because I chose to work around lack of some primitive operations rather than bloat Delta adding them).

**Comment by tomjaguarpaw** [comment]
> I can't easily relate the infinite case to the ML convolution

The same results should hold for ML convolution-with-zero-padding, that is

```
conv(x, y) . z = x . conv(y, z)
conv(x, y) = conv(y, x)
dr = conv(x, dy) + conv(dx, y) = conv(x, dy) + conv(y, dx)
∂/∂y = conv(∂/∂r, x)
∂/∂x = conv(∂/∂r, y) = conv(y, ∂/∂r)
```

**Comment by Mikolaj** [comment]
You mean, we can think of the padding as being infinite in both directions? Makes sense. What is not yet clear to me is, e.g., how to obtain ∂/∂r, assuming I want to compute ∂/∂y and I'm given x, y, dx and dy.

BTW, it now occurred to me, we probably only need

```Haskell
conv2' :: IsScalar r
       => DualNumber (Tensor2 r) -> HM.Matrix r -> DualNumber (Tensor2 r)
conv2' ker@(D u _) v = D (HM.conv2 u v) ???
```

because we are only interested in the gradient of the kernel, not of the big matrix, which holds the input data. We train by tweaking the shared weights in the kernel (and a bias, but that's trivial and can be added on top), not by modifying the data we are supposed to recognize.

So, we are only interested in, say, ∂/∂y from your exposition and we can assume ∂/∂x is zero (or something like that).

**Comment by tomjaguarpaw** [comment]
> You mean, we can think of the padding as being infinite in both directions?

Yes exactly.

> What is not yet clear to me is, e.g., how to obtain ∂/∂r, assuming I want to compute ∂/∂y and I'm given x, y, dx and dy.

` ∂/∂r` is the `r` argument to [`eval2`](https://github.com/Mikolaj/horde-ad/blob/748b805ed9917915e20a266264d48d04c0fa47e9/src/HordeAd/Internal/Delta.hs#L441=). (`dx` and `dy` are not relevant to computing ` ∂/∂y`.

> we are only interested in the gradient of the kernel, not of the big matrix

Yes, that's right. We could restrict ourselves to "convolution of a constant image by a variable kernel".

> we are only interested in, say, ∂/∂y from your exposition and we can assume ∂/∂x is zero

Yes indeed.

**Comment by Mikolaj** [comment]
> > What is not yet clear to me is, e.g., how to obtain ∂/∂r, assuming I want to compute ∂/∂y and I'm given x, y, dx and dy.
> 
> ` ∂/∂r` is the `r` argument to [`eval2`](https://github.com/Mikolaj/horde-ad/blob/748b805ed9917915e20a266264d48d04c0fa47e9/src/HordeAd/Internal/Delta.hs#L441=). (`dx` and `dy` are not relevant to computing ` ∂/∂y`.

That confuses me. Let me rephrase the original signature of `conv2'`:

```Haskell
conv2' :: DualMonad r m
       => DualNumber (Tensor2 r) -> DualNumber (Tensor2 r)
       -> m (DualNumber (Tensor2 r))
conv2' (D x dx) (D y dy) = D (HM.conv2 x y) ???
```

Do I understand it right that ??? won't refer to dx nor dy?

**Comment by tomjaguarpaw** [comment]
> Do I understand it right that ??? won't refer to dx nor dy?

No, it *will* refer to them, but it will not refer to `∂/∂r`.

```haskell
conv2' :: DualMonad r m
       => DualNumber (Tensor2 r) -> DualNumber (Tensor2 r)
       -> m (DualNumber (Tensor2 r))
conv2' (D x dx) (D y dy) = D (HM.conv2 x y) (convish x dy + convish dx y)
```

where `convish` 2-d convolution with the correct type signature. (I'm not sure how your types line up.) The form below is the gradient map form, only relevant to `eval2`.

```
∂/∂y = conv(∂/∂r, x)
∂/∂x = conv(∂/∂r, y) = conv(y, ∂/∂r)
```

**Comment by Mikolaj** [comment]
Hah, OK, so I was asking "how to express the derivative using simpler operations that we know how to evaluate", while you observations were answering "how to evaluate (that is, compute gradient based on) convishA/B". If that works, that's brilliantly simple.

**Comment by Mikolaj** [comment]
I've implemented and pushed the most naive convolutional neutral network for MNIST and it appears to train, but is hopelessly slow. So now would be the right time to try the fused variant of @conv2@, check performance, QuickCheck the neural net with it has the same derivative and gradient as with the slow @conv2@, etc. 

@tomjaguarpaw: have you made any experiments with the brilliantly simple derivative for convolution? If so, could you push or share? I'm going to experiment soon, but I'd hate to duplicate your work. BTW, does correlation, especially without padding or with limited padding, have any similarly nice properties? You are right ML (often?) uses correlation, because the kernels are randomly initialized so it really doesn't matter whether they are reversed or not.

**Comment by tomjaguarpaw** [comment]
No, I haven't written any code for this.

Correlation with padding can be give a nice derivative by observing that `corr(x, y) = conv(x^R, y)`.

Correlation/convolution without padding are just correlation/convolution with padding followed by a trimming operation, so to proceed you just need the derivative of the trimming operation which is ... the trimming operation!  (And you'll need its gradient for `eval2`, which is zero padding).

**Comment by Mikolaj** [comment]
So, assuming I have

```Haskell
conv2' :: IsScalar r
       => DualNumber (Tensor2 r) -> DualNumber (Tensor2 r)
       -> DualNumber (Tensor2 r)
conv2' (D u u') (D v v') = D (HM.conv2 u v) (Conv2 u v' + Conv2 v u')
```

and
 ```Haskell
 | Conv2 (Matrix r) (Delta2 r)
```

and

```Haskell
evalBindingsForward.eval2 parameters (Conv2 m md) == HM.conv2 m $ eval2 parameters md
```

How do I use

> ∂/∂r is the r argument to eval2
> ∂/∂y = conv(∂/∂r, x)
> ∂/∂x = conv(∂/∂r, y) = conv(y, ∂/∂r)

to derive the code for `buildFinMaps.eval2 r (Conv2 m md)`? I imagine it would contain `eval2 ??? md`, but that's the only inkling I have.

**Comment by tomjaguarpaw** [comment]
I think it is

```haskell
eval r (Conv2 m md) = eval (conv2 m r) md
```

**Comment by tomjaguarpaw** [comment]
(It follows the same pattern as scale and dot)

**Comment by Mikolaj** [comment]
QuickCheck indeed mostly confirms (hard to tell for sure, because it's so slow and some of the equalities are fuzzy due to floating point error)

https://github.com/Mikolaj/horde-ad/blob/50a16046dc387d6aa1a4655b785277059fc372c6/test/TestMnistCNN.hs#L380-L383

your formula here

https://github.com/Mikolaj/horde-ad/blob/50a16046dc387d6aa1a4655b785277059fc372c6/src/HordeAd/Core/DualNumber.hs#L632

but balks 

https://github.com/Mikolaj/horde-ad/blob/50a16046dc387d6aa1a4655b785277059fc372c6/test/TestMnistCNN.hs#L384-L386

at this attempt

https://github.com/Mikolaj/horde-ad/blob/50a16046dc387d6aa1a4655b785277059fc372c6/src/HordeAd/Internal/Delta.hs#L497-L501

**Comment by tomjaguarpaw** [comment]
Hmm yes, I suppose my formula only works for infinite sizes. I will think about how to handle the situation of finite sizes.

**Comment by tomjaguarpaw** [comment]
> QuickCheck indeed mostly confirms (hard to tell for sure, because it's so slow

That's strange. Why would it be slow? Do you test large randomly-generated programs? If so why not also test single functions?

**Comment by tomjaguarpaw** [comment]
> I will think about how to handle the situation of finite sizes.

I think the answer is that you have to pad the kernel and perhaps trim the image, but I'll try to come up with a more precise answer.

**Comment by Mikolaj** [comment]
> That's strange. Why would it be slow? Do you test large randomly-generated programs? If so why not also test single functions?

I test the whole convolutional nn for MNIST, with two nested calls to `conv2` (and `conv2'` in the other version and I compare the two). I don't test randomly generated programs with `conv2` against the same programs with `conv2'`, though this would be great to have. Testing 'single-functions', that is, `conv2` against `conv2'` with only variables (i.e., parameters) as arguments would catch this error, too, but in general would be very weak, because for `conv2'` that would exercise depth 2 delta-expressions only, for which evaluation may potentially do something wrong that masks the problem (e.g., `Index0` applied to a variable is evaluated specially).

This is so slow, because `conv2` is slow, because it goes down to the level of scalar deltas, losing all unboxing and creating huge delta expressions. That was expected, though I hoped it's going to be fast enough at least for QuickCheck tests against the fast variant. Perhaps it can be improved a bit to make this viable and, e.g., the test can be turned on when running CI.

**Comment by tomjaguarpaw** [comment]
> TODO: sizes of moc (2 * m + md - 2) and md don't match 

I am pretty sure the correct answer is to trim `moc` so its size matches `m`, and it should be done symmetrically, that is chop `m - 1` off both the top and bottom (resp. left and right).

Please give that a try. I think there's a 90% chance it works fine. In the 10% case please report back and I'll have another go.

**Comment by Mikolaj** [comment]
It's usually close (the outer cells don't matter that much), but still fails early

https://github.com/Mikolaj/horde-ad/tree/trim-conv-eval

Perhaps pad the little one instead?

**Comment by tomjaguarpaw** [comment]
It's somewhat puzzling, but this is fairly convincing that the correct formula is

```
∂/∂y = conv(∂/∂r^R, x)^R
∂/∂x = conv(∂/∂r^R, y)^R = conv(y, ∂/∂r^R)^R
```

where `^R` is rotation of the matrix by 180 degrees.  In your code, if you replace `convolved = HM.conv2 m mor` by `convolved = rotate (HM.conv2 m (rotate mor))` then I expect it will work.

```haskell
{-# OPTIONS_GHC -Wall #-}

module Main where

import Numeric.LinearAlgebra

x :: Matrix Double
x = fromLists [[5, 6, 7, 8]
              , [12, 9, 10, 11]
              ]

dx :: Matrix Double
dx = fromLists [[15, 30, 8, 5]
               , [1, 2, 3, 4]
               ]

y :: Matrix Double
y = fromLists [[20, 21],
               [22, 23],
               [25, 30]]

dy :: Matrix Double
dy = fromLists [[1, 9],
                [7, 6 ],
                [2, 3 ]]

dr :: Matrix Double
dr = conv2 x dy + conv2 dx y

d_dr :: Matrix Double
d_dr = fromLists [[1, 2, 3, 4, 5],
                  [6, 7, 8, 9, 10],
                  [-1, -2, 16, 5, 2],
                  [5, 4, -6, -10, 20]]

d_dy :: Matrix Double
d_dy = let d_dy1 = rotate (conv2 x (rotate d_dr))
           (r, c) = size y
           (r', c') = size d_dy1
       in subMatrix ((r' - r) `div` 2, (c' - c) `div` 2) (r, c) d_dy1

d_dx :: Matrix Double
d_dx = let d_dx1 = rotate (conv2 (rotate d_dr) y)
           (r, c) = size x
           (r', c') = size d_dx1
       in subMatrix ((r' - r) `div` 2, (c' - c) `div` 2) (r, c) d_dx1

dotM :: Matrix Double -> Matrix Double -> Double
dotM m1 m2 = dot (flatten m1) (flatten m2)

example :: IO ()
example = do
  print ((d_dx `dotM` dx) + (d_dy `dotM` dy) )
  print (d_dr `dotM` dr)

reflectH :: Matrix Double -> Matrix Double
reflectH = Prelude.foldl1 (===)
           . Prelude.reverse
           . Prelude.map asRow
           . toRows

reflectV :: Matrix Double -> Matrix Double
reflectV = Prelude.foldl1 (|||)
           . Prelude.reverse
           . Prelude.map asColumn
           . toColumns

rotate :: Matrix Double -> Matrix Double
rotate = reflectV . reflectH
```

**Comment by tomjaguarpaw** [comment]
Actually you can do it more simply:

```
convolved = HM.conv2 (rotate m) mor
```

**Comment by Mikolaj** [comment]
So is the gradient of convolution a correlation or something?

One round of QuickCheck tests passed fine. Yay. Pushed to https://github.com/Mikolaj/horde-ad/tree/trim-conv-eval

**Comment by tomjaguarpaw** [comment]
> So is the gradient of convolution a correlation or something?

Yes, looks like it. I'll have to go back through my algebra to discover where I made an error, and correct it.

**Comment by Mikolaj** [comment]
WIth this tests pass fine, too, so I'm going to merge this variant to master:

```Haskell
        Conv2 m md ->
          let mor = MO.convertMatrixOuter r
              convolved = HM.corr2 m mor
              moc = MO.MatrixOuter (Just convolved) Nothing Nothing
          in eval2 moc md
```

**Comment by tomjaguarpaw** [comment]
That's great!

**Comment by tomjaguarpaw** [comment]
So I guess that correlation is already the rotation-plus-trimming operation on top of convolution.

**Comment by Mikolaj** [comment]
Yes. Or, inversely (as it's implemented in hmatrix), convolution is correlation plus rotation plus padding: https://github.com/Mikolaj/horde-ad/blob/891dfa54a8fe63006dec681db748d0d0afd4ca29/src/HordeAd/Core/DualNumber.hs#L627

**Comment by tomjaguarpaw** [comment]
> > So is the gradient of convolution a correlation or something?
> 
> Yes, looks like it. I'll have to go back through my algebra to discover where I made an error, and correct it.

I corrected my algebra above.

**Comment by Mikolaj** [comment]
> I corrected my algebra above.

This algebra, I presume? 

https://github.com/Mikolaj/horde-ad/issues/24#issuecomment-1079915637

I have to say this is quite beautiful. 

In other news, the latest version using your Delta constructor `Conv2`, now pushed to master, is twice faster than the version built from primitive dual number functions (with the disclaimer that our benchmarks are currently worth nought, due to #14 and #20). We could halve the runtime again by using correlation instead of convolution and as little padding as needed not to shrink the matrix, with a variant of `Conv2` as its derivative. But this is still incredibly slow, so I think we'd need to use the `im2col` trick or the trick from `tf.nn.conv2d`, if different, to get away with just a single matrix multiplication instead of so many dot products. Again, the dual component of the dual number will be something similar to `Conv2`, but we'd also need to evaluate it using `im2col` or it's going to become the bottleneck. To be continued in #9 and let's open a new ticket about defining and evaluating its delta-expression once the semantics of the primal part of `im2col/tf.nn.conv2d` is clear (e.g., whether it's a convolution or correlation, what padding/trimming, etc.).

Shall we declare success and close?

**Comment by tomjaguarpaw** [comment]
> This algebra, I presume?

Yes, exactly.

**Comment by tomjaguarpaw** [comment]
> Shall we declare success and close?

Yes, I think so. We have the operation correctly implemented.

**Comment by Mikolaj** [comment]
> If so why not also test single functions?

We are now testing (with QuickCheck) the single conv2 functions as well --- delta-expression evaluations currently has no special case speedups for matrix variables, so this is very close to testing raw output of the conv2 functions. Tests pass so far.

**Comment by Mikolaj** [comment]
I just remembered the QuickCheck tests (which I verified to be fine and cover rectangle matrices) work fine also with convolution, trimmed afterwards

https://github.com/Mikolaj/horde-ad/commit/df2319a80daca6e20354168d34499a66c19cbd77#diff-f52cfd8e6458258571edd4888ee9c4bbd5f92909f5ca748f94c349a4d09b489eL499

which likely means the trimmed bits of the result are just zeros and I need to trim them only because dot product and others expect vectors/matrices with matching sizes. So your derivation is correct and the equations are true in the infinite case.

(By happy coincidence, the trimming that is needed is exactly the one that not padding at all induces and hmatrix.corr2 does not pad at all (as opposed to hmatrix.conv2 that pads maximally). BTW, the tiny padding called SAME, that is only enough to ensure input matrix size does not decrease, already makes the MNIST computations twice faster.)

I'm completely satisfied now and I will add code comments that match your derivation (from the google doc and form here) and that include the `eval` invariant you have been using. BTW, how should I pronounce, name and relate to other quantities (e.g., to `dr`) the ∂/∂r value that is equal to the scaling factor `r` from `eval`?
